From f0b1d8566c5a54b4f28e080abaab4e3c0ec69e4e Mon Sep 17 00:00:00 2001
From: nhatthieu <nhat.thieu.xr@renesas.com>
Date: Tue, 19 Apr 2022 09:20:39 +0700
Subject: [PATCH 6/6] Workaround GPU driver remove power domains v2l

Signed-off-by: nhatthieu <nhat.thieu.xr@renesas.com>
---
 arch/arm64/boot/dts/renesas/r9a07g054.dtsi | 1 -
 1 file changed, 1 deletion(-)

diff --git a/arch/arm64/boot/dts/renesas/r9a07g054.dtsi b/arch/arm64/boot/dts/renesas/r9a07g054.dtsi
index ba75c5a3710a..7d18e3f1aa72 100644
--- a/arch/arm64/boot/dts/renesas/r9a07g054.dtsi
+++ b/arch/arm64/boot/dts/renesas/r9a07g054.dtsi
@@ -1182,7 +1182,6 @@ gpu: gpu@11840000 {
 				 <&cpg CPG_MOD R9A07G054_GPU_AXI_CLK>,
 				 <&cpg CPG_MOD R9A07G054_GPU_ACE_CLK>;
 			clock-names = "gpu", "bus", "bus_ace";
-			power-domains = <&cpg>;
 			resets = <&cpg R9A07G054_GPU_RESETN>,
 				 <&cpg R9A07G054_GPU_AXI_RESETN>,
 				 <&cpg R9A07G054_GPU_ACE_RESETN>;
-- 
2.34.1

